src/toxcore/messenger/packet/msi.rs

```diff
@@ -0,0 +1,303 @@
+        2 => value!(MsiErrorKind::InvalidParam) |
+        3 => value!(MsiErrorKind::InvalidState) |
+        4 => value!(MsiErrorKind::StrayMessage) |
+        5 => value!(MsiErrorKind::System) |
+        6 => value!(MsiErrorKind::Handle) |
+        7 => value!(MsiErrorKind::Undisclosed)
+    ));
+}
+
+#[derive(Clone, Copy, Debug, Eq, PartialEq)]
+enum MsiSubPacket {
+    Request(Request),
+    MsiError(MsiError),
+    Capabilities(Capabilities),
+}
+
+impl FromBytes for MsiSubPacket {
+    named!(from_bytes<MsiSubPacket>, alt!(
+        map!(Request::from_bytes, MsiSubPacket::Request) |
+        map!(MsiError::from_bytes, MsiSubPacket::MsiError) |
+        map!(Capabilities::from_bytes, MsiSubPacket::Capabilities)
+    ));
+}
+
+impl ToBytes for MsiSubPacket {
+    fn to_bytes<'a>(&self, buf: (&'a mut [u8], usize)) -> Result<(&'a mut [u8], usize), GenError> {
+        match *self {
+            MsiSubPacket::Request(ref p) => p.to_bytes(buf),
+            MsiSubPacket::MsiError(ref p) => p.to_bytes(buf),
+            MsiSubPacket::Capabilities(ref p) => p.to_bytes(buf),
+        }
+    }
+}
+
+/** Msi is a struct that holds info for Media Session Interface.
+
+Sub-packet: kind [1 byte], size [1 byte], value [$size bytes] : but actually size is always 1, so a sub-packet is always 3 bytes long
+
+- kind: one of Request, Capabilities, Error
+- size: the length in byte of value(always 1)
+- value: enum value depending on kind
+
+Payload: |sub_packet| |...{sub-packet}| |0|
+
+Serialized form:
+
+Length    | Content
+--------- | -------
+`1`       | `0x45`
+`0..255`  | payload
+
+Sub-packet serialized form:
+
+Length    | Content
+--------- | -------
+`1`       | kind(1 = Request, 2 = Error, 3 = Capabilities)
+`1`       | size(always 1)
+`1`       | value(it depends on kind)
+
+*/
+#[derive(Clone, Debug, Eq, PartialEq)]
+pub struct Msi {
+    request: Request,
+    error: Option<MsiError>,
+    capabilities: Capabilities,
+}
+
+impl Msi {
+    /// Make new msi struct
+    pub fn new(request: RequestKind, error: Option<MsiErrorKind>, capabilities: CapabilitiesKind) -> Self {
+        Msi {
+            request: Request(request),
+            error: error.map(|error| MsiError(error)),
+            capabilities: Capabilities(capabilities),
+        }
+    }
+
+    fn remove_redundant(input: &[u8], sub_packets: Vec<MsiSubPacket>) -> IResult<&[u8], Msi> {
+        let mut request = None;
+        let mut error = None;
+        let mut capabilities = None;
+        for sub in sub_packets.clone() {
+            match sub {
+                MsiSubPacket::Request(req) => { request = Some(req); },
+                MsiSubPacket::MsiError(err) => { error = Some(err); },
+                MsiSubPacket::Capabilities(capa) => { capabilities = Some(capa); },
+            };
+        }
+        let msi = Msi {
+            request: request.expect(&format!("Missing request sub-packet in msi packet: {:?}", sub_packets)),
```

Comment:
Replace expect with returning nom error, otherwise our client will be easy to kill :)

Reply:
done